text/0085-api-keys.md

```diff
@@ -0,0 +1,463 @@
+
+If the value of the main key changes, the previously generated `API Keys` do not change or expire. Thus, MeiliSearch does not re-generate the previous generated default API keys.
+
+If the main key is removed at MeiliSearch launch, the previously generated API keys no longer secure the MeiliSearch instance.
+
+If MeiliSearch is launched with the `production` value for the `MEILI_ENV` environment variable or the `--env` CLI option, a main key is mandatory to force the user to secure his instance. If the main key is omitted in that particular case, MeiliSearch launch is aborted and displays the `Error: In production mode, the environment variable MEILI_MAIN_KEY is mandatory` error in stdout.
+
+> Note that the main key does not appear on the `/keys` endpoints.
+
+---
+
+#### `API Key` object representation
+
+| field       | type    | description                                      |
+|-------------|---------|--------------------------------------------------|
+| description | string  | A description for the key. `null` if empty. |
+| key         | string  | The generated key. **Generated by MeiliSearch**. |
+| actions     | array   | A list of actions permitted for the key. `["*"]` for all actions. See Actions list definition part. |
+| indexes     | array   | A list of indexes permitted for the key. `["*"]` for all indexes. |
+| expiresAt   | string  | Represent the expiration date and time as `ISO-8601` format. `null` equals to no expiration time. |
+| createdAt   | string  | Represent the date and time as `ISO-8601` format when the API key has been created. **Generated by MeiliSearch** |
+| updatedAt   | string  | Represent the date and time as `ISO-8601` format when the API key has been updated. **Generated by MeiliSearch** |
+
+#### `POST`/ `PATCH` - `/keys` - API Key object payload definition
+
+| field       | type    | required |description                     |
+|-------------|---------|----------|--------------------------------|
+| indexes     | array   | Required | `[*]` for all indexes. **Default**: `No Default` |
+| description | string  | Optional | A description for the API key. **Default**: `null` |
+| actions     | array   | Required | A list of actions permitted for the API key. `["*"]` for all actions. **See Actions list definition part**. `*` character can be used as a wildcard. e.g. `documents.*` to authorize access on all documents endpoints. **Default**: `No default` |
+| expiresAt   | string  | Required | The expiration date and time as `ISO-8601` format. `null` equals to no expiration time. Sending only the date part e.g `2021-12-01` leads to having an `expiresAt` value set to `2021-12-01T00:00:00`. **Default**: `No Default` |
```

Comment:
> Why make it mandatory and allow users to set it to `null` anyway?

Mostly for security concerns. By making it mandatory, it force the user to think of it and explicitly set that the API Key being generated will not have an expiration date by passing `null` instead of omitting the field.

Reply:
> Also, are there any best/recommended practices regarding `expiresAt`? E.g.: avoid creating keys that never expire, avoid creating many keys with short expiry dates, etc

Very good question @guimachiavelli. 

In general, it is a good practice to always put expiration dates for security reasons. The fact of not putting any is in the eyes of the user who must understand the importance of his decision.

For a company, it also allows to have governance on the management of the keys and to make sure that the rotation of the keys is done at regular intervals. Regularly scheduled changes to an API secret ensure that their app is less vulnerable to being compromised.